Fixes

p-steuer opened the pull request:
icsf, tpm and ep11 still need to be tested.
ica, cca looking good.
soft: tdes hmac still fails.

Fixes https://github.com/opencryptoki/opencryptoki/issues/285.

I know its a lot of code duplication, but the <= OpenSSL 1.0.2 code could be removed, soon: https://github.com/opencryptoki/opencryptoki/issues/293

jschmidb commented:
tpm_openssl.c line 134 ff:
>     if (pkey != NULL)
>         EVP_PKEY_free(pkey);
Like the other "free" routines, EVP_PKEY_free does the NULL check internally. I know, we currently almost always do this check before calling a free-routine, but let's think about removing such unnecessary checks.

p-steuer commented:
Good point, i ll remove those checks.

p-steuer commented:
Moving this out opf WIP. Successfully tested soft, ica and cca. Soft tdes cmac still fails with openssl master, but it seems to be a problem with openssl and i submitted a fix. Its working with the fix. Still needs to be reviewed and tested with icsf, ep11 and tpm.

ifranzki commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -224,6 +224,7 @@ CK_RV token_specific_des_ecb(STDLL_TokData_t *tokdata,
                              CK_ULONG *out_data_len,
                              OBJECT *key, CK_BYTE encrypt)
 {
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
```
patch title: `Soft: replace dedicated OpenSSL interfaces`
I guess you mean **deprecated**, not **dedicated** ....

p-steuer replied:
done

ifranzki commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -2970,13 +3629,132 @@ CK_RV token_specific_tdes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
+            goto err;
+        }
+
+        cmac->pkey = EVP_PKEY_new_CMAC_key(NULL,
```
I get compile errors when compiling on my system:

```
usr/lib/soft_stdll/.libs/opencryptoki_stdll_libpkcs11_sw_la-soft_specific.o: In function `token_specific_tdes_cmac':
/root/opencryptoki-pr292/usr/lib/soft_stdll/soft_specific.c:3691: undefined reference to `EVP_PKEY_new_CMAC_key'
usr/lib/soft_stdll/.libs/opencryptoki_stdll_libpkcs11_sw_la-soft_specific.o: In function `token_specific_aes_cmac':
/root/opencryptoki-pr292/usr/lib/soft_stdll/soft_specific.c:3883: undefined reference to `EVP_PKEY_new_CMAC_key'
```
with this OpenSSL version:
```
# openssl version
OpenSSL 1.1.0h-fips  27 Mar 2018
```
I guess EVP_PKEY_new_CMAC_key only exists since OpenSSL 1.1.1.
So I guess the OpenSSL version check in this function is not correct:
`#if OPENSSL_VERSION_NUMBER < 0x10100000L`

p-steuer replied:
yes, should check for 1.1.1

p-steuer replied:
done

ifranzki commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -3050,4 +3828,115 @@ CK_RV token_specific_aes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
+            goto err;
+        }
+
+        cmac->pkey = EVP_PKEY_new_CMAC_key(NULL,
```
same here

p-steuer replied:
done

ifranzki commented on usr/lib/tpm_stdll/tpm_openssl.c:
```diff
@@ -73,27 +75,40 @@ RSA *openssl_gen_key()
     }
 
-    rc = RSA_check_key(rsa);
+    rc = EVP_PKEY_check(ctx) == 1 ? 1 : 0;
```
EVP_PKEY_check is also only available since OpenSSL 1.1.1.

The #ifdef OLDER_OPENSSL checks for the wrong OpenSSL version for this specific function:

```
#if OPENSSL_VERSION_NUMBER < 0x10100000L
#define OLDER_OPENSSL
#endif
```

p-steuer replied:
yes. i should not use the already existing OLDER_OPENSSL macro since it checks for 1.1.0 vs 1.0.2. I should use a new 1.1.1 macro for all of my changes since the apis are not deprectaed in 1.1.0 anyway so theres no need to change that.

p-steuer replied:
done

ifranzki commented:
EP11 token looks good (with slightly adjusted code to get it compiled with OpenSSL 1.1.0).

ifranzki commented:
ICSF token shows `C_Login() rc = CKR_FUNCTION_FAILED` on all testcases.
Trace:
```
04/15/2020 13:52:58 17076 [usr/lib/api/api_interface.c:2990 api] INFO: C_OpenSession  10 6 (nil) (nil) 0x3ffdb57bba8
04/15/2020 13:52:58 17076 [usr/lib/icsf_stdll/new_host.c:577 icsftok] INFO: C_OpenSession: rc = 0x00000000 sess = 1
04/15/2020 13:52:58 17076 [usr/lib/api/api_interface.c:3032 api] DEVEL: fcn->ST_OpenSession returned: 0x0
04/15/2020 13:52:58 17076 [usr/lib/api/api_interface.c:2873 api] INFO: C_Login
04/15/2020 13:52:58 17076 [usr/lib/api/api_interface.c:2892 api] INFO: Valid Session handle id: 1
04/15/2020 13:52:58 17076 [usr/lib/icsf_stdll/pbkdf.c:137 icsftok] ERROR: EVP_DecryptFinal failed.
04/15/2020 13:52:58 17076 [usr/lib/icsf_stdll/icsf_specific.c:1272 icsftok] DEVEL: Failed to load master key.
04/15/2020 13:52:58 17076 [usr/lib/icsf_stdll/new_host.c:890 icsftok] INFO: C_Login: rc = 0x00000006
04/15/2020 13:52:58 17076 [usr/lib/api/api_interface.c:2906 api] DEVEL: fcn->ST_Login returned:0x6
```
This is in `decrypt_aes()` in pkbdf.c. 
Don't know why this fails, since only the context creation is different with your changes.

p-steuer commented:
I cant imagine it being due to context creation. Also it occurs at master key encryption, not at racf encryption where the same aes functions are used. I think it must be due to the pbkdf2 changes. pbkdf2 is used to derive the key which (un)wrapps the masterkey.

ifranzki commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2572,6 +2573,53 @@ CK_RV token_specific_des_ecb(STDLL_TokData_t * tokdata,
+        return CKR_FUNCTION_FAILED;
+    }
+
+    memcpy(dkey, attr->pValue, sizeof(dkey));
```
RSA testcases on TPM token segfaults here (line 2592):

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000003ff94564412 in token_specific_des_ecb (tokdata=<optimized out>,
    in_data=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352",
    in_data_len=32, out_data=0x3ffe3cfe010 "", out_data_len=out_data_len@entry=0x3ffe3cfdf50, key=0x293d9d10, encrypt=0 '\000')
    at usr/lib/tpm_stdll/tpm_specific.c:2592
2592        memcpy(dkey, attr->pValue, sizeof(dkey));
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.27-32.1.ibm.fc28.s390x
(gdb) bt
#0  0x000003ff94564412 in token_specific_des_ecb (tokdata=<optimized out>,
    in_data=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352",
    in_data_len=32, out_data=0x3ffe3cfe010 "", out_data_len=out_data_len@entry=0x3ffe3cfdf50, key=0x293d9d10, encrypt=0 '\000')
    at usr/lib/tpm_stdll/tpm_specific.c:2592
#1  0x000003ff94553f5c in ckm_des_ecb_decrypt (tokdata=<optimized out>, in_data=<optimized out>, in_data_len=<optimized out>,
    out_data=<optimized out>, out_data_len=0x3ffe3cfdf50, key=0x293d9d10) at usr/lib/common/mech_des.c:1464
#2  0x000003ff945540e6 in des_ecb_decrypt (tokdata=0x28ff3c20, sess=<optimized out>, length_only=<optimized out>,
    ctx=<optimized out>,
    in_data=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352",
    in_data_len=32, out_data=0x3ffe3cfe010 "", out_data_len=0x3ffe3cfdf50) at usr/lib/common/mech_des.c:127
#3  0x000003ff9454b34a in decr_mgr_decrypt (tokdata=0x28ff3c20, sess=0x290ed0f0, length_only=<optimized out>, ctx=0x290ed170,
    in_data=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352",
    in_data_len=32, out_data=0x3ffe3cfe010 "", out_data_len=0x3ffe3cfdf50) at usr/lib/common/decr_mgr.c:694
#4  0x000003ff9451e65e in SC_Decrypt (tokdata=0x28ff3c20, sSession=sSession@entry=0x3ffe3cfde20,
    pEncryptedData=pEncryptedData@entry=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352", ulEncryptedDataLen=ulEncryptedDataLen@entry=32, pData=pData@entry=0x3ffe3cfe010 "", pulDataLen=0x3ffe3cfdf50)
    at usr/lib/common/new_host.c:2324
#5  0x000003ff9528433e in C_Decrypt (hSession=<optimized out>,
    pEncryptedData=0x3ffe3cfdff0 "\210\065^\202v\250f\371\314\301P\005\201\342\203\302\334\220p\352;\033\003INW\241D\326\370\272\352", ulEncryptedDataLen=32, pData=0x3ffe3cfe010 "", pulDataLen=0x3ffe3cfdf50) at usr/lib/api/api_interface.c:395
#6  0x0000000001009d2c in do_WrapUnwrapRSA (tsuite=tsuite@entry=0x10384e0 <generated_keywrap_test_suites>)
    at testcases/crypto/rsa_func.c:1315
#7  0x000000000100c7f8 in rsa_funcs () at testcases/crypto/rsa_func.c:1752
#8  0x0000000001000dfe in main (argc=<optimized out>, argv=<optimized out>) at testcases/crypto/rsa_func.c:1805
```
attr->pValue is NULL:
`*attr = {type = 17, pValue = 0x0, ulValueLen = 0}`

This is the same error as mentioned here: https://github.com/opencryptoki/opencryptoki/issues/280#issuecomment-600105191
So although its in your modified code, ist still he same old error, not cuased by your changes.

However, AES, DES, 3DES tests run fine.

ifranzki commented:
I guess your patch "Correct soft token model name everywhere …" should be integrated soon, since Travis tests won't run any tests if it does not detect that the soft token is there.

p-steuer commented:
I think all patches would be ready to go except the icsf / pbkdf change.

p-steuer commented:
Adressed all comments and rebased.

I moved the ICSF PBKDF commit that still has problems to a separate PR (https://github.com/opencryptoki/opencryptoki/pull/318) so this one can be included in the next release.

juergenchrist commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+
+    // Generate the DH Key
+    if (!DH_generate_key(dh)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
```
Shouldn't we free dh, bn_p, and bn_g here?

p-steuer replied:
done

juergenchrist commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
```
Shouldn't we free dh, bn_p, bn_g, and temp_byte here?

p-steuer replied:
done

juergenchrist commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+                         prime_attr->ulValueLen, &temp_attr);  // in bytes
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
         return CKR_FUNCTION_FAILED;
```
Shouldn't we free dh, bn_p, bn_g, and value_bits_attr here?

p-steuer replied:
Done for dh, bn_p, bn_g. The template was already updated at this place to point to value_bits_attr, so its applications responsibility to free the template in case of an error.

juergenchrist commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+                         base_attr->ulValueLen, &temp_attr);     // in bytes
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        return CKR_FUNCTION_FAILED;
```
Shouldn't we free dh, bn_p, bn_g, and value_bits_attr here?

p-steuer replied:
Done for dh, bn_p, bn_g. The template was already updated at this place to point to value_bits_attr, so its applications responsibility to free the template in case of an error.

hfreude commented on usr/lib/common/template.c:
```diff
@@ -638,25 +636,25 @@ CK_RV template_unflatten_withSize(TEMPLATE **new_tmpl, CK_BYTE *buf,
 
                 {
-                    CK_ULONG_32 *p32;
+                    CK_ULONG_32 p32;
```
another variable name like u32 would be better, the p is probably a leftover from pointer

p-steuer replied:
done

hfreude commented on usr/lib/common/template.c:
```diff
@@ -638,25 +636,25 @@ CK_RV template_unflatten_withSize(TEMPLATE **new_tmpl, CK_BYTE *buf,
-                    attr_ulong_32 = *p32;
+                    memcpy(&p32, pb2, sizeof(p32));
+                    attr_ulong_32 = p32;
                 }
```
can't we do all the {} above much better ?
attr_ulong_32 = *((CK_ULONG_32 *)(((CK_BYPTE *)(ptr)) + sizeof(CK_ATTRIBUTE_32)));
lot of parenthesis ...

p-steuer replied:
No, you cant cast an u8 pointer (alignment 1) to an u32 (alignment 4) and dereference it without address sanitizer complaining.

juergenchrist replied:
What about endianness here?

hfreude commented on usr/lib/common/template.c:
```diff
@@ -701,7 +699,7 @@ CK_RV template_unflatten_withSize(TEMPLATE **new_tmpl, CK_BYTE *buf,
         if (long_len == 4)
```
that's ugly, mark it with /* NEEDS REWORK */ or similar
or find out what exactly happens here

p-steuer replied:
It tries to determine at runtime if we are running on a 32 bit (4byte long type) or 64 bit (8byte long type) platform. At least for z this works. We do it at more places, you can open an issue for that if you want, but i thunk we can remove 32 bit support anyway soon, together with the old openssl code. We probably should do both for the next release.

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2570,6 +2571,53 @@ CK_RV token_specific_des_ecb(STDLL_TokData_t * tokdata,
+
+    memcpy(dkey, attr->pValue, sizeof(dkey));
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
is there a define or constant for the  des block size of 8 ?

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2619,6 +2668,53 @@ CK_RV token_specific_des_cbc(STDLL_TokData_t * tokdata,
+
+    memcpy(dkey, attr->pValue, sizeof(dkey));
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
same here, use some define or for the 8 byte DES block size.

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2619,6 +2668,53 @@ CK_RV token_specific_des_cbc(STDLL_TokData_t * tokdata,
+    const EVP_CIPHER *cipher = EVP_des_cbc();
+    EVP_CIPHER_CTX *ctx = NULL;
+    CK_ATTRIBUTE *attr = NULL;
+    unsigned char dkey[8];
```
dkey[DES_KEY_SIZE]

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2702,6 +2799,67 @@ CK_RV token_specific_tdes_ecb(STDLL_TokData_t * tokdata,
+        memcpy(dkey, attr->pValue, 3 * DES_KEY_SIZE);
+    }
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
use some kind of DES_BLOCK_SIZE

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -2783,6 +2942,67 @@ CK_RV token_specific_tdes_cbc(STDLL_TokData_t * tokdata,
+        memcpy(dkey, attr->pValue, 3 * DES_KEY_SIZE);
+    }
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
use constant instead of 8 and move this check before the key value is copied into the key buffer

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -3513,6 +3734,63 @@ CK_RV token_specific_aes_ecb(STDLL_TokData_t * tokdata,
+#else
+    CK_RV rc;
+    int outlen;
+    unsigned char akey[32];
```
do we have a constant for AES_256_KEY_SIZE ?

p-steuer replied:
yes we do. done.

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -3513,6 +3734,63 @@ CK_RV token_specific_aes_ecb(STDLL_TokData_t * tokdata,
+    else if (keylen == 192 / 8)
+        cipher = EVP_aes_192_ecb();
+    else if (keylen == 256 / 8)
+        cipher = EVP_aes_256_ecb();
```
else TRACE_ERROR("key len mismatch ?!?!") ...
or simple
else if (keylen == 256 / 8)    ->  else

p-steuer replied:
done

hfreude commented on usr/lib/tpm_stdll/tpm_specific.c:
```diff
@@ -3551,6 +3830,63 @@ CK_RV token_specific_aes_cbc(STDLL_TokData_t * tokdata,
+        cipher = EVP_aes_192_cbc();
+    else if (keylen == 256 / 8)
+        cipher = EVP_aes_256_cbc();
+
```
else ?

p-steuer replied:
done

hfreude commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -273,6 +274,53 @@ CK_RV token_specific_des_ecb(STDLL_TokData_t *tokdata,
+
+    memcpy(dkey, attr->pValue, sizeof(dkey));
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
move this if before the memcpy
as you my return here without cleaning the key buffer

p-steuer replied:
done

hfreude commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -320,6 +369,53 @@ CK_RV token_specific_des_cbc(STDLL_TokData_t *tokdata,
+
+    memcpy(dkey, attr->pValue, sizeof(dkey));
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
move this block befoe the memcpy of the key

p-steuer replied:
done

hfreude commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -320,6 +369,53 @@ CK_RV token_specific_des_cbc(STDLL_TokData_t *tokdata,
+    const EVP_CIPHER *cipher = EVP_des_cbc();
+    EVP_CIPHER_CTX *ctx = NULL;
+    CK_ATTRIBUTE *attr = NULL;
+    unsigned char dkey[8];
```
dkey[DES_KEY_SIZE]

p-steuer replied:
done

hfreude commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -401,6 +498,67 @@ CK_RV token_specific_tdes_ecb(STDLL_TokData_t *tokdata,
+        memcpy(dkey, attr->pValue, 3 * DES_KEY_SIZE);
+    }
+
+    if (in_data_len % 8 || in_data_len > INT_MAX) {
```
move before the key copy

p-steuer replied:
done

hfreude commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -589,6 +809,12 @@ static void *rsa_convert_private_key(OBJECT *key_obj)
         meth->bn_mod_exp = meth2->bn_mod_exp;
 #else
+/*
+ * XXX I dont see a better was than to ignore this warning for now.
```
was -> way

p-steuer replied:
done

p-steuer commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1001,13 +1263,23 @@ static CK_RV os_specific_rsa_keygen(TEMPLATE *publ_tmpl, TEMPLATE *priv_tmpl)
+    if (e != NULL)
+        BN_free(e);
+    if (ssl_ptr != NULL)
+        free(ssl_ptr);
```
done

p-steuer commented on usr/lib/soft_stdll/soft_specific.c:
```diff
@@ -1951,42 +2465,172 @@ CK_RV token_specific_dh_pkcs_derive(STDLL_TokData_t *tokdata,
+    // in bytes
+    rc = build_attribute(CKA_VALUE, temp_byte, temp_bn_len, &temp_attr);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
```
done

p-steuer commented:
Thanks Harald and Jürgen for the reviews. I addressed all of your findings. If Traavis is green and noone objects, im going to merge this later.
